### PR TITLE
google-vertexai[minor]: added safety_settings property to gemini wrapper

### LINKED ISSUE #14841: safety_settings argument is missing with the ChatVertexAI mode
### Issue you'd like to raise.

safety_settings argument is missing with the `ChatVertexAI` model

### Suggestion:

Should be able to define the safety settings. For example:

```
safety_settings = {
  HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_NONE,
  HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_NONE,
  HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_NONE,
  HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,
}

llm = ChatVertexAI(safety_settings=safety_settings)
```

### PR DESCRIPTION
**Pull request by izapolsk**
**Description:** Gemini model has quite annoying default safety_settings settings. In addition, current VertexAI class doesn't provide a property to override such settings.
So, this PR aims to 
 - add safety_settings property to VertexAI
- fix issue with incorrect LLM output parsing when LLM responds with appropriate 'blocked' response
- fix issue with incorrect parsing LLM output when Gemini API blocks prompt itself as inappropriate
- add safety_settings related tests

I'm not enough familiar with langchain code base and guidelines. So, any comments and/or suggestions are very welcome.
 
**Issue:** it will likely fix #14841

### CODE REVIEW

**Comment by izapolsk** [comment]
@efriis , @hwchase17, could you please give you approval to re-run workflows ?  I suppose those have to run w/o issues this time. Thanks in advance !

**Comment by izapolsk** [comment]
@efriis, could you please review the PR ?

**Comment by nicsuzor** [comment]
Thanks for this, I also hit this problem! I've tested your code and it works for me. You may want to also make these changes in chat_models/vertexai.py as well -- see e.g. [add-safety_settings-to-chat_models.vertexai.patch](https://github.com/langchain-ai/langchain/files/13881892/0001-add-safety_settings-to-chat_models.vertexai.patch)

**Comment by baron** [comment]
I ended up ripping out the Langchain version because of this. Would be nice if we could at least pass any option we want using kwargs or such. The other thing is this thin wrapper simply fails and halts the entire chain without any feedback. You guys were quick to launch the Google Gemini project with practically no follow through and there's no real way to escalate important issues other than commenting on GitHub at regular intervals.

**Review by efriis — COMMENTED** [review]
Great addition! I'll actually move this over to the `langchain-google-vertexai` package instead (and add it to chat model if relevant), as these changes are made on the deprecated community implementations for vertex.

to use the new features, would recommend switching over!

https://python.langchain.com/docs/integrations/llms/google_vertex_ai_palm

**Comment by nicsuzor** [comment]
Nice! I was just looking at patching the `langchain-google-vertexai` package as you posted this. In case its useful, I also wrote a docstring entry for `class ChatGoogleGenerativeAI(BaseChatModel):`
```

    safety_settings: Optional[dict[Any, Any]] = None
    """The default safety settings to use for all generations. 
    
        For example: 

            from vertexai.preview.generative_models import HarmBlockThreshold, HarmCategory

            safety_settings = {
                HarmCategory.HARM_CATEGORY_UNSPECIFIED: HarmBlockThreshold.BLOCK_NONE,
                HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_ONLY_HIGH,
                HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_LOW_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE,
            }
            """
```

And add `"safety_settings": self.safety_settings,` to [_prepare_params](https://github.com/langchain-ai/langchain/blob/ebb6ad4f7acf63692415bd48098fdfb1ad81125e/libs/partners/google-genai/langchain_google_genai/chat_models.py#L522)

**Comment by efriis** [comment]
@nicsuzor amazing. Will include it!

I'll probably leave this out of the community package, because we don't have the same guarantees on the sdk version people are using there, so it would be a breaking change to have the same typing. Hope that's ok!

**Comment by efriis** [comment]
Hey @nicsuzor @izapolsk @lkuligin ! Curious what you think of this state. Haven't finished documentation for `llm` yet, and there were a few redundant passing-ins of `safety_settings` that I *think* are better handled like this because it's only supported on gemini models. Let me know what you think!

**Comment by izapolsk** [comment]
> Hey @nicsuzor @izapolsk @lkuligin ! Curious what you think of this state. Haven't finished documentation for `llm` yet, and there were a few redundant passing-ins of `safety_settings` that I _think_ are better handled like this because it's only supported on gemini models. Let me know what you think!

nice, LGTM. only thing is I'd leave tests as is. those were covering more cases: - blocked prompt, blocked response, etc

File: libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py [hunk 1]
```diff
@@ -212,9 +213,16 @@ class ChatVertexAI(_VertexAICommon, BaseChatModel):
...
+        safety_settings = values["safety_settings"]
+
+        if safety_settings and not is_gemini:
+            raise ValueError("Safety settings are only supported for Gemini models")
+
         cls._init_vertexai(values)
         if is_gemini:
-            values["client"] = GenerativeModel(model_name=values["model_name"])
+            values["client"] = GenerativeModel(
+                model_name=values["model_name"], safety_settings=safety_settings
```
Comment (lkuligin) [inline libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py:224]: `safety_filters` can be also passed to `send_message` (and override the initial config) , e.g. here: https://github.com/langchain-ai/langchain/blob/5ca74c784eee0a76c0e22a0d293d9a793fb0f1c9/libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py#L274

so I'd suggest to allow passing `safety_filters` through `kwargs` to `_generate` (and use them :) )

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py [hunk 2]
```diff
@@ -227,8 +253,12 @@ def validate_environment(cls, values: Dict) -> Dict:
             )
         else:
             if is_gemini:
-                values["client"] = model_cls(model_name=model_name)
-                values["client_preview"] = preview_model_cls(model_name=model_name)
+                values["client"] = model_cls(
```
Comment (lkuligin) [inline libs/partners/google-vertexai/langchain_google_vertexai/llms.py:256]: same as for chat_models, allow passing `safety_filters` through kwargs to `_generate`

**Comment by lkuligin** [comment]
added some comments, but LGTM in general, thanks for adding this!

File: libs/partners/google-vertexai/pyproject.toml [hunk 3]
```diff
@@ -8,7 +8,7 @@ readme = "README.md"
 [tool.poetry.dependencies]
 python = ">=3.8.1,<4.0"
 langchain-core = ">=0.1,<0.2"
-google-cloud-aiplatform = "1.38.1"
+google-cloud-aiplatform = "^1.39.0"
```
Comment (lkuligin) [inline libs/partners/google-vertexai/pyproject.toml:11]: why do we need to bump the version?

File: libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py (hunk 1 above)
Comment (izapolsk) [reply libs/partners/google-vertexai/langchain_google_vertexai/chat_models.py:224]: added handling safety_settings in _generate/_agenerate. 
it'd been present in initial commit but it disappeared after some changes.

File: libs/partners/google-vertexai/langchain_google_vertexai/llms.py (hunk 2 above)
Comment (izapolsk) [reply libs/partners/google-vertexai/langchain_google_vertexai/llms.py:259]: done

**Comment by izapolsk** [comment]
@efriis , @lkuligin , please review

**Review by lkuligin — APPROVED** [review]

**Comment by lkuligin** [comment]
Thanks a lot for your contribution!

**Review by efriis — APPROVED** [review]

**Comment by baron** [comment]
@efriis are there any plans to integrate this into the Google GenAI wrapper as well?

#15095

**Comment by blackslashcreative** [comment]
I saw that this was added to the docs but when I try to use safety settings with ChatVertexAI I get KeyError: 'safety_settings'
Screenshot 2024-02-24 at 12 21 07 PM
Just trying a simple example from the docs, but it doesn't work: 
`chat = ChatVertexAI(
    model_name="gemini-pro",
    safety_settings={
      HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_NONE
    },
)`

**Comment by JorgePardoSerrano** [comment]
Just in case anyone is still having problems with safety settings and Gemini: You don't define them when you initialize your LLM. Instead, you have to send the safety settings with each invoke operation.

To make it work, you have to invoke the chain using "with_config".

```
safety_settings = {
    HarmCategory.HARM_CATEGORY_UNSPECIFIED: HarmBlockThreshold.BLOCK_ONLY_HIGH,
    HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
    HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_ONLY_HIGH,
    HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
    HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_ONLY_HIGH
}
```

Create your chain and invoke it like this:

`chain.with_config(configurable={"safety_settings": safety_settings}).invoke('<your prompt>')`

If you are not using a chain then just do this:

```
model_name = "gemini-1.5-flash-001"
llm = ChatVertexAI(
    model=model_name,
    convert_system_message_to_human=False,
    handle_parsing_errors=True,
    temperature=0.7,
    max_tokens=2048,
)
llm.invoke("<your prompt>", safety_settings=safety_settings)
```

**Comment by peerdavid** [comment]
> Just in case anyone is still having problems with safety settings and Gemini: You don't define them when you initialize your LLM. Instead, you have to send the safety settings with each invoke operation.
> 
> To make it work, you have to invoke the chain using "with_config".
> 
> ```
> safety_settings = {
>     HarmCategory.HARM_CATEGORY_UNSPECIFIED: HarmBlockThreshold.BLOCK_ONLY_HIGH,
>     HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
>     HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_ONLY_HIGH,
>     HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
>     HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_ONLY_HIGH
> }
> ```
> 
> Create your chain and invoke it like this:
> 
> `chain.with_config(configurable={"safety_settings": safety_settings}).invoke('<your prompt>')`
> 
> If you are not using a chain then just do this:
> 
> ```
> model_name = "gemini-1.5-flash-001"
> llm = ChatVertexAI(
>     model=model_name,
>     convert_system_message_to_human=False,
>     handle_parsing_errors=True,
>     temperature=0.7,
>     max_tokens=2048,
> )
> llm.invoke("<your prompt>", safety_settings=safety_settings)
> ```

Then I highly recommend to update the documentation

```
ChatVertexAI

Key init args — completion params:

[...]

safety_settings: Optional[Dict[vertexai.generative_models.HarmCategory, vertexai.generative_models.HarmBlockThreshold]]
The default safety settings to use for all generations.
```

As well as the code example given:

```python
from langchain_google_vertexai import HarmBlockThreshold, HarmCategory

llm = ChatVertexAI(
    model="gemini-1.5-pro",
    safety_settings={
        HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_LOW_AND_ABOVE,
        HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
        HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_LOW_AND_ABOVE,
        HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_ONLY_HIGH,
    },
)

llm.invoke(messages).response_metadata
```

From https://api.python.langchain.com/en/latest/chat_models/langchain_google_vertexai.chat_models.ChatVertexAI.html#langchain-google-vertexai-chat-models-chatvertexai

**Pull request merged**